multichain-aggregator/multichain-aggregator-logic/src/page_token.rs

```diff
@@ -0,0 +1,186 @@
+                }
+            }
+        )*
+    };
+}
+
+impl_page_token_format!(i32, i64, Address, B256, String);
+
+impl PageTokenFormat for Decimal {
+    fn parse_page_token(page_token: String) -> Result<Self, PageTokenParsingError> {
+        page_token
+            .parse()
+            .map_err(|_| PageTokenParsingError::ParsingError(page_token, 0))
+    }
+
+    fn format_page_token(self) -> String {
+        self.normalize().to_string()
+    }
+}
+
+impl PageTokenFormat for BigDecimal {
+    fn parse_page_token(page_token: String) -> Result<Self, PageTokenParsingError> {
+        page_token
+            .parse()
+            .map_err(|_| PageTokenParsingError::ParsingError(page_token, 0))
+    }
+
+    fn format_page_token(self) -> String {
+        self.to_plain_string()
+    }
+}
+
+impl PageTokenFormat for NaiveDateTime {
+    fn parse_page_token(page_token: String) -> Result<Self, PageTokenParsingError> {
+        match page_token
+            .parse()
+            .map(chrono::DateTime::from_timestamp_micros)
+        {
+            Ok(Some(dt)) => Ok(dt.naive_utc()),
+            _ => Err(PageTokenParsingError::ParsingError(page_token, 0)),
+        }
+    }
+
+    fn format_page_token(self) -> String {
+        self.and_utc().timestamp_micros().to_string()
+    }
+}
+
+// Tuples are formatted as base64-encoded JSON array of strings
+// where each element is formatted as a string using the PageTokenFormat trait
+macro_rules! impl_page_token_format_tuple {
+    ($len:literal: ($($T:ident $var:ident),+)) => {
+        impl<$($T: PageTokenFormat),+> PageTokenFormat for ($($T),+,) {
+            fn parse_page_token(page_token: String) -> Result<Self, PageTokenParsingError> {
+                let decoded = base64::engine::general_purpose::URL_SAFE_NO_PAD
+                    .decode(page_token)
+                    .map_err(|e| PageTokenParsingError::Base64DecodeError(e.to_string()))?;
+
+                let parts: Vec<&str> = serde_json::from_slice(&decoded)
+                    .map_err(|e| PageTokenParsingError::JsonDecodeError(e.to_string()))?;
+
+                match parts.as_slice() {
+                    &[$($var),+] => Ok((
+                        $(
+                            $T::parse_page_token($var.to_string())?,
+                        )+
+                    )),
+                    _ => Err(PageTokenParsingError::FormatError($len as u32)),
+                }
+            }
+
+            fn format_page_token(self) -> String {
+                let ($($var),+,) = self;
+                let parts = vec![
+                    $(
+                        $var.format_page_token(),
+                    )+
+                ];
+                let json = serde_json::to_string(&parts).expect("list of strings should be serializable");
+                base64::engine::general_purpose::URL_SAFE_NO_PAD.encode(json)
+            }
+        }
+    };
+}
+
+impl_page_token_format_tuple!(1: (T1 v1));
```

Comment:
we can get rid of len parameter with something like this

https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=55eb689fb15ef46f369d53db06b86ee7

(adapted from https://stackoverflow.com/questions/76567540/rust-macro-count-repetitions-of-nested-match)